server/src/main/java/com/microsoft/java/bs/core/internal/services/LifecycleService.java

```diff
@@ -119,61 +134,120 @@ public void exit() {
+      if (gradleJavaHome != null) {
+
+        try {
+          String gradleJavaHomeVersion = JavaUtils.getJavaVersionFromFile(gradleJavaHome);
+          if (
+              JavaUtils.isCompatible(
+                  gradleJavaHomeVersion,
+                  oldestCompatibleVersion,
+                  latestCompatibleVersion
+              )
+          ) {
+            return gradleJavaHome;
+          }
+        } catch (IOException e) {
+          LOGGER.severe("Invalid GradleJavaHome: " + e.getMessage());
+        }
+
       }
 
-      if (StringUtils.isNotBlank(gradleVersion)) {
-        map = TelemetryUtils.getMetadataMap("gradleVersion", gradleVersion);
-        LOGGER.log(Level.INFO, "Gradle version: " + gradleVersion, map);
-        String highestJavaVersion = Utils.getHighestCompatibleJavaVersion(gradleVersion);
-        File jdkInstallation = getJdkToLaunchDaemon(preferences.getJdks(), highestJavaVersion);
-        if (jdkInstallation != null) {
-          preferences.setGradleJavaHome(jdkInstallation.getAbsolutePath());
+      // Pick a compatible JDK from the JDKs available in Preferences
+      if (preferences.getJdks() != null && preferences.getJdks().isEmpty()) {
+        File selectedJdk = getLatestCompatibleJdk(
+            preferences.getJdks(),
+            oldestCompatibleVersion,
+            latestCompatibleVersion
+        );
+        if (selectedJdk != null) {
+          return selectedJdk;
         }
       }
+
+    }
+
+    // Notify client, for no compatible JDK can be found
+    if (client != null) {
+      ShowMessageParams messageParams = new ShowMessageParams(
+          MessageType.ERROR,
+          "Failed to find a JDK compatible with current gradle version "
+              + "(" + gradleVersion + "). Compatible JDK versions include ("
+              + oldestCompatibleVersion + " - " + latestCompatibleVersion + ")"
+      );
+      client.onBuildShowMessage(messageParams);
     }
+
+    return null;
+
   }
 
   /**
-   * Find the latest JDK but equal or lower than the {@code highestJavaVersion}.
+   * Finds the latest version of JDK from the given map of jdks
+   * between {@code oldestCompatibleJavaVersion} and {@code latestCompatibleJavaVersion}.
    */
-  static File getJdkToLaunchDaemon(Map<String, String> jdks, String highestJavaVersion) {
-    if (StringUtils.isBlank(highestJavaVersion)) {
-      return null;
-    }
+  static File getLatestCompatibleJdk(
+      Map<String, String> jdks,
+      String oldestCompatibleJavaVersion,
+      String latestCompatibleJavaVersion
+  ) {
 
     Entry<String, String> selected = null;
     for (Entry<String, String> jdk : jdks.entrySet()) {
       String javaVersion = jdk.getKey();
-      if (Version.parse(javaVersion).compareTo(Version.parse(highestJavaVersion)) <= 0
-          && (selected == null || Version.parse(selected.getKey())
-              .compareTo(Version.parse(javaVersion)) < 0)) {
+      Boolean isSelectedVersionHigher =
```

Comment:
nit:

1. There is no need to use object Boolean here, just primitive `boolean` should be fine.
2. Seems that name the variable to `isHigherThanSelected` makes more sense.

Reply:
Addressed in commit [e7de6d8](https://github.com/microsoft/build-server-for-gradle/pull/165/commits/e7de6d870c1893f46e4d25f1bfb2b59ffed35218)